### PR TITLE
rbd: add functions/constants related to rbd pool stat options

### PR DESCRIPTION
**Pull request by agarwal-mudit**
Added constants for rbd pool stat options along with wrappers for following functions:
1. rbd_pool_init()
2. rbd_pool_stats_create()
3. rbd_pool_stats_destroy()
4. rbd_pool_stats_option_add_uint64)
5. rbd_pool_stats_get()

Fixes: https://github.com/ceph/go-ceph/issues/278

Signed-off-by: Mudit Agarwal <muagarwa@redhat.com>

## Checklist
- [ ] Added tests for features and functional changes
- [ ] Public functions and types are documented
- [ ] Standard formatting is applied to Go code

### CODE REVIEW

File: rbd/pool_nautilus_test.go [hunk 1]
```diff
@@ -99,3 +99,30 @@ func TestPoolMetadata(t *testing.T) {
...
+	ioctx, err := conn.OpenIOContext(poolName)
+	assert.NoError(t, err)
+	assert.NotNil(t, ioctx)
+
+	defer func() {
+		ioctx.Destroy()
+		conn.DeletePool(poolName)
+		conn.Shutdown()
+	}()
+
```
Comment (nixpanic) [inline rbd/pool_nautilus_test.go:119]: Can you add a test with incorrect parameters, like `ioctx == nil`?

File: rbd/pool_nautilus.go [hunk 2]
```diff
@@ -98,3 +98,27 @@ func PoolInit(ioctx *rados.IOContext, force bool) error {
 	ret := C.rbd_pool_init(cephIoctx(ioctx), C.bool(force))
 	return getError(ret)
 }
+
+// PoolStats represents RBD pool stats variable.
+type PoolStats struct {
+	stats C.rbd_pool_stats_t
+}
+
+// PoolStatsCreate a new PoolStats struct.
```
Comment (nixpanic) [inline rbd/pool_nautilus.go:107]: missing the word 'creates'

File: rbd/pool_nautilus.go [hunk 3]
```diff
@@ -98,3 +98,27 @@ func PoolInit(ioctx *rados.IOContext, force bool) error {
...
+	C.rbd_pool_stats_create(&poolStats.stats)
+	return poolStats
+}
+
+// Destroy a PoolStats struct and free the associated resources.
+//
+// Implements:
+//  void rbd_pool_stats_destroy(rbd_pool_stats_t stats)
+func (poolStats *PoolStats) Destroy() {
+	C.rbd_pool_stats_destroy(poolStats.stats)
```
Comment (nixpanic) [inline rbd/pool_nautilus.go:123]: consider setting `poolStats.stats` to `nil` to prevent problematic usage after destroying.

File: rbd/pool_nautilus.go [hunk 4]
```diff
@@ -122,3 +122,61 @@ func PoolStatsCreate() *PoolStats {
 func (poolStats *PoolStats) Destroy() {
 	C.rbd_pool_stats_destroy(poolStats.stats)
 }
+
+// PoolStatOptions represents a group of configurable pool stat options.
+type PoolStatOptions C.rbd_pool_stat_option_t
```
Comment (nixpanic) [inline rbd/pool_nautilus.go:127]: Is this a single option? Call it `PoolStatOption`.

File: rbd/pool_nautilus.go [hunk 5]
```diff
@@ -122,3 +122,61 @@ func PoolStatsCreate() *PoolStats {
...
+	PoolStatOptionTrashSnapshots = PoolStatOptions(C.RBD_POOL_STAT_OPTION_TRASH_SNAPSHOTS)
+)
+
+// AddOption adds the given PoolStatOptions to PoolStats.
+// Pass the address of a variable where the value of the option needs to be stored.
+//
+// Implements:
+//  int rbd_pool_stats_option_add_uint64(rbd_pool_stats_t stats, int stat_option, uint64_t* stat_val)
+func (poolStats *PoolStats) AddOption(option PoolStatOptions, val *uint64) error {
+
```
Comment (nixpanic) [inline rbd/pool_nautilus.go:162]: why an empty line?

File: rbd/pool_nautilus.go [hunk 6]
```diff
@@ -122,3 +122,61 @@ func PoolStatsCreate() *PoolStats {
...
+	// RBD_POOL_STAT_OPTION_TRASH_MAX_PROVISIONED_BYTES from librbd.
+	PoolStatOptionTrashMaxProvisionedBytes = PoolStatOptions(C.RBD_POOL_STAT_OPTION_TRASH_MAX_PROVISIONED_BYTES)
+	// PoolStatOptionTrashSnapshots is the representation of
+	// RBD_POOL_STAT_OPTION_TRASH_SNAPSHOTS from librbd.
+	PoolStatOptionTrashSnapshots = PoolStatOptions(C.RBD_POOL_STAT_OPTION_TRASH_SNAPSHOTS)
+)
+
+// AddOption adds the given PoolStatOptions to PoolStats.
+// Pass the address of a variable where the value of the option needs to be stored.
+//
```
Comment (nixpanic) [inline rbd/pool_nautilus.go:158]: Maybe mention that after calling `AddOption()`, calling `Get()` can be used to retrieve the values for the options.

File: rbd/pool_nautilus.go (hunk 2 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:107]: Thanks, added.

File: rbd/pool_nautilus.go (hunk 5 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:162]: Fixed

File: rbd/pool_nautilus.go (hunk 6 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:158]: Fixed

File: rbd/pool_nautilus.go (hunk 4 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:127]: Fixed

File: rbd/pool_nautilus_test.go (hunk 1 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus_test.go:119]: Added

File: rbd/pool_nautilus.go (hunk 3 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:123]: Added

Comment (nixpanic) [reply rbd/pool_nautilus.go:123]: now I wonder what happens when you call `poolStat.Destroy()` 2x :wink:

Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:123]: I guess that's fine, I have a testcase where I am calling poolStats.Destroy()

Comment (nixpanic) [reply rbd/pool_nautilus.go:123]: did you try calling it twice? some of the Ceph library functions crash when getting called with a `null` argument.

Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:123]: But why some one should call Destroy() twice? Its like calling free() twice

Comment (nixpanic) [reply rbd/pool_nautilus.go:123]: We should not, but users could. Adding a check for `poolStats.stats != nil` makes it less likely that users get a segfault on incorrect API usage.

File: rbd/pool_nautilus.go [hunk 7]
```diff
@@ -84,3 +84,17 @@ func RemovePoolMetadata(ioctx *rados.IOContext, key string) error {
 	ret := C.rbd_pool_metadata_remove(cephIoctx(ioctx), cKey)
 	return getError(ret)
 }
+
+// PoolInit is a stub for handling new pool initialization.
+// It currently only handles setting the application tag to rbd.
```
[the remark this answers is not in the document]
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:89]: Thanks for the suggestion, for the lack of better words I copied what was mentioned in cep docs :)
Changed it now.

File: rbd/pool_nautilus.go [hunk 8]
```diff
@@ -98,3 +98,28 @@ func PoolInit(ioctx *rados.IOContext, force bool) error {
...
+type PoolStats struct {
+	stats C.rbd_pool_stats_t
+}
+
+// PoolStatsCreate creates a new PoolStats struct.
+// Call PoolStats.Destroy to free the resources.
+//
+// Implements:
+//  void rbd_pool_stats_create(rbd_pool_stats_t *stats)
+func PoolStatsCreate() *PoolStats {
```
[the remark this answers is not in the document]
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:112]: Fixed

File: rbd/pool_nautilus.go (hunk 3 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:123]: Added a check

File: rbd/pool_nautilus.go [hunk 9]
```diff
@@ -123,3 +123,61 @@ func (poolStats *PoolStats) Destroy() {
...
+// addresses passed in AddOption.
+//
+// Implements:
+//  int rbd_pool_stats_get(rados_ioctx_t io, rbd_pool_stats_t stats)
+func (poolStats *PoolStats) Get(ioctx *rados.IOContext) error {
+	if ioctx == nil {
+		return ErrNoIOContext
+	}
+
+	ret := C.rbd_pool_stats_get(cephIoctx(ioctx), (poolStats.stats))
```
[the remark this answers is not in the document]
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:181]: Fixed

**Comment by agarwal-mudit** [comment]
> I didn't dig into the tests much once I realized there was a bit of a mismatch between the Golang-way and the C api ceph uses here. Generally, we try to provide a thin wrapper around the C apis of Ceph, but try to make them feel Go idomatic as much as (reasonably) possible. Sometimes an api is awkward enough that I think we need to make a judgement call and deviate a little from what is happening at the C layer. As long as we don't place extra limits on the power of the API I think this is generally a good thing.
> 
> With generalities out of the way, the specific issue I see here is that the C api expects you to set up the "fetcher" with bunch of rbd_pool_stats_option_add_uint64 calls that take pointers to ints, and then call rbd_pool_stats_get which will fetch the stats and fill in pointer values. This is slightly similar to what I was also working on in PR #315 but a bit simpler and more homogeneous. Because the return types are the same I propose a simplified call, partially inspired by the how this function is wrapped by ceph's python code (in https://github.com/ceph/ceph/blob/octopus/src/pybind/rbd/rbd.pyx#L2556 ).
> 
> ```go
> // GetPoolStats returns a map of requested PoolStatOptions to their values. 
> func GetPoolStats(ioctx *rados.IOContext, o []PoolStatOption) (map[PoolStatOption]uint64, error) {
>     // .... implementation ....
> }
> ```
> 
> Unlike the python code, the user specifies what options they want, which is more similar to the C api. If we wanted to, perhaps in the future, we could even do a convenience version of the function that gets all known PoolStatOptions, just like the python code.
> 
> Now, I'm not dead set on this suggested API but I do think we need to pause and consider the API design for this subset of librados. We do want to keep our layer thin, but not too thin. Let's discuss. :-)

So, AFAICT we have two options:

1. We don't expose AddOption() API and set all the options during GetPoolStats() only as done in the python code.

2. We expose AddOption() but ask user to only set the option not the pointer to the option value but internally define a map (of options and valeues) and based on that map passess the address of the respective pointer to the option value inside AddOption(). 

Something like:

```
createPooStatMap() {  // this can be created and destroyed during CreatePoolStats() and Destroy() if required.
  m[PoolStatOptionImages] = cImages                                                                     
  m[PoolStatOptionImageProvisionedBytes] = cImageProvisionBytes 
  m[PoolStatOptionImageMaxProvisionedBytes = cImageProvisionMaxBytes
}

AddOption(PoolStatOption option) {
      optionVal = m[option]
      C.rbd_pool_stats_option_add_uint64(poolStata.stats, C.in(option), &optionVal)
}

GetOption() {
// define another struct for GetOption, not the same map as in AddOption()
return &poolStats {
                cImages: cImages
                cImageProvisionBytes: cImageProvisionBytes
		cImageProvisionMaxBytes: cImageProvisionMaxBytes
	}

```
PS: Just a thought, haven't looked from implementation point of view.

In both the cases, GetPoolStats() will return the map of all the values as done in the python code.

**Comment by phlogistonjohn** [comment]
I certainly lean toward option 1, as my suggested api might indicate. I feel that this api is the way it is due to some of C's quirks and since Go is much more oriented toward more & richer data types, and I generally assume people using the API want something simple to just get the stats they need, there isn't a great impetus to make the API in go-ceph as granular as the one exposed in C.

**Comment by agarwal-mudit** [comment]
I have un exported AddOption() and added two APIs:

1. func (poolStats *PoolStats) GetPoolStats(ioctx *rados.IOContext, option PoolStatOption) (uint64, error)
>> Returns value for a single Pool Stat Option

2. func GetAllPoolStats(ioctx *rados.IOContext) (map[PoolStatOption]uint64, error)
>> Returns a map of all Pool Stat Options to their respective values.

Have pushed the changes to  https://github.com/agarwal-mudit/go-ceph/tree/pool but somehow that is not updating this PR automatically :(

File: rbd/pool_nautilus.go [hunk 10]
```diff
@@ -84,3 +84,151 @@ func RemovePoolMetadata(ioctx *rados.IOContext, key string) error {
...
+	if ioctx == nil {
+		return ErrNoIOContext
+	}
+
+	ret := C.rbd_pool_init(cephIoctx(ioctx), C.bool(force))
+	return getError(ret)
+}
+
+// PoolStats represents RBD pool stats variable.
+type PoolStats struct {
```
Comment (phlogistonjohn) [inline rbd/pool_nautilus.go:104]: I'm not sure what the value of exporting this as a type is. It seem unnecessary to me.

File: rbd/pool_nautilus.go [hunk 11]
```diff
@@ -84,3 +84,151 @@ func RemovePoolMetadata(ioctx *rados.IOContext, key string) error {
...
+	if ioctx == nil {
+		return omap, ErrNoIOContext
+	}
+
+	poolStats := PoolStatsCreate()
+	defer func() {
+		poolStats.Destroy()
+	}()
+
+	var keys [8]PoolStatOption
```
Comment (phlogistonjohn) [inline rbd/pool_nautilus.go:204]: I don't understand the approach used here. Why not set up the map in the first place like:
```
m := map[PoolStatOption]uint64{
    PoolStatOptionImages: 0,
    PoolStatOPtionImageProvsionedBytes: 0,
    // ...and so on...
}
```

File: rbd/pool_nautilus.go [hunk 12]
```diff
@@ -84,3 +84,151 @@ func RemovePoolMetadata(ioctx *rados.IOContext, key string) error {
...
+func (poolStats *PoolStats) addPoolStatOption(option PoolStatOption, val *uint64) error {
+	ret := C.rbd_pool_stats_option_add_uint64(
+		poolStats.stats,
+		C.int(option),
+		(*C.uint64_t)(val))
+	return getError(ret)
+}
+
+// GetPoolStats returns the value for the PoolStatOption.
+func (poolStats *PoolStats) GetPoolStats(ioctx *rados.IOContext, option PoolStatOption) (uint64, error) {
```
Comment (phlogistonjohn) [inline rbd/pool_nautilus.go:174]: Interesting, but like the type being exported this seem like it might not provide a lot of value to the caller, versus just using GetAllPoolStats and just grabbing the one value you care about out of the map. I'd argue that we should skip this for now but if there's demand for it at a later time (say some sort of measurable performance difference) we can add it later.

File: rbd/pool_nautilus.go (hunk 11 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:204]: I need to pass the address in addOption() and I can't pass the address of map value, so either I pass the address through different variables same like python code or I keep a slice for values and pass the address of the respective values so that value slice is named ovalArray.

Now, why I used a slice named keys is because I need to iterate in an order but during map iteration order of keys is not guranteed. I coded the suggested approach first but changed it because of the order issue.

File: rbd/pool_nautilus.go (hunk 12 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:174]: Agree, will remove it. Wanted to get an opinion based on the last conversation.

File: rbd/pool_nautilus.go (hunk 10 above)
Comment (agarwal-mudit) [reply rbd/pool_nautilus.go:104]: We can unexport it if we are just going with GetAllPoolStats() and in that case we should not even export PoolStatsCreate() and Destroy() because those two are also only used by GetAllPoolStats()

Comment (phlogistonjohn) [reply rbd/pool_nautilus.go]: Agreed.

File: rbd/pool_nautilus.go (hunk 11 above)
Comment (phlogistonjohn) [reply rbd/pool_nautilus.go:187]: OK, that's a bit annoying but I understand now.

**Comment by agarwal-mudit** [comment]
@nixpanic @phlogistonjohn PTAL

File: rbd/pool_nautilus.go [hunk 13]
```diff
@@ -99,3 +99,30 @@ func PoolInit(ioctx *rados.IOContext, force bool) error {
...
+// poolStats represents RBD pool stats variable.
+type poolStats struct {
+	stats C.rbd_pool_stats_t
+}
+
+// poolStatsCreate creates a new poolStats struct.
+//
+// Implements:
+//  void rbd_pool_stats_create(rbd_pool_stats_t *stats)
+func poolStatsCreate() *poolStats {
```
Comment (phlogistonjohn) [inline rbd/pool_nautilus.go:112]: Since these are unexported functions we could just return the C types now. However it's not a big deal at all.

File: rbd/pool_nautilus_test.go [hunk 14]
```diff
@@ -131,3 +132,80 @@ func TestPoolInit(t *testing.T) {
...
+			assert.NoError(t, err)
+			assert.NoError(t, TrashRemove(ioctx, trashList[0].Id, true))
+		}()
+
+		omap, err := GetAllPoolStats(ioctx)
+		assert.NoError(t, err)
+
+		assert.Equal(t, uint64(3), omap[PoolStatOptionImages])
+		assert.Equal(t, (expectedSize - size), omap[PoolStatOptionImageProvisionedBytes])
+		assert.Equal(t, expectedSize-size, omap[PoolStatOptionImageMaxProvisionedBytes])
```
Comment (phlogistonjohn) [inline rbd/pool_nautilus_test.go:204]: No biggie, by why the parens on the line above and none here?

Comment (agarwal-mudit) [reply rbd/pool_nautilus_test.go:204]: Not intentional, just an accident.
